rebound/examples/PureSystemF.hs

```diff
@@ -0,0 +1,280 @@
+    TAbs bnd -> TAbs (r bnd)
+    TApp t1 t2 -> TApp (r t1) (r t2)
+    where
+      r :: forall t. (Subst Exp t) => t n -> t m
+      r = applyE env
+
+-- We will be needing strengthening in the type-checker;
+-- more on that later.
+instance Strengthen Exp where
+  strengthenRec ::
+    forall k m n.
+    SNat k ->
+    SNat m ->
+    SNat n ->
+    Exp (k + (m + n)) ->
+    Maybe (Exp (k + n))
+  strengthenRec k m n t = case t of
+    Var x -> Var <$> strengthenRec k m n x
+    Kind -> return Kind
+    TAll bnd -> TAll <$> r bnd
+    TArr t1 t2 -> TArr <$> r t1 <*> r t2
+    Abs ty bnd -> Abs <$> r ty <*> r bnd
+    App t1 t2 -> App <$> r t1 <*> r t2
+    TAbs bnd -> TAbs <$> r bnd
+    TApp t1 t2 -> TApp <$> r t1 <*> r t2
+    where
+      r :: (Strengthen t) => t (k + (m + n)) -> Maybe (t (k + n))
+      r = strengthenRec k m n
+
+--------------------------------------------------------------------------------
+--- Typechecking
+--------------------------------------------------------------------------------
+
+-- | An environment mapping (de Bruijn) variables to
+-- a user-defined name and its type.
+data TcEnv n = TcEnv
+  { names :: Vec n LocalName,
+    types :: Ctx Exp n
+  }
+
+emptyEnv :: TcEnv Z
+emptyEnv = TcEnv {names = Vec.empty, types = zeroE}
+
+-- | Add a new binding to the environment
+extendE :: (LocalName, Exp n) -> TcEnv n -> TcEnv (S n)
+extendE (n, t) (TcEnv ns ts) =
+  TcEnv (n ::: ns) (ts +++ t)
+
+-- | Search for a binding. Lookup cannot fail
+-- thanks to extrinsic scoping.
+lookupE :: TcEnv n -> Fin n -> (LocalName, Exp n)
+lookupE (TcEnv ns ts) i = (ns ! i, applyEnv ts i)
+
+type Error = String
+
+-- | Typechecking monad.
+newtype TC n a = TC (ScopedReaderT TcEnv (Except Error) n a)
+  deriving (Functor, Applicative, Monad, MonadError Error)
+
+-- Trivial lifting through a newtype.
+instance MonadScopedReader TcEnv TC where
+  askS = TC askS
+  localS f (TC m) = TC (localS f m)
+
+-- | Run the type-checking monad. Returns
+-- either the result, or an error.
+runTC :: TcEnv n -> TC n a -> Either Error a
+runTC env (TC m) = runExcept $ runScopedReaderT m env
+
+-- | Extend the current (latent) scope with a new binding.
+push :: LocalName -> Exp n -> TC (S n) a -> TC n a
+push n t = Scoped.localS $ extendE (n, t)
+
+-- | Lookup a binding in the (latent) scope.
+get :: Fin n -> TC n (LocalName, Exp n)
+get i = readerS (`lookupE` i)
+
+-- | Checks that a given type is indeed a (valid) type,
+-- by ensuring that its own type is 'Kind'.
+ensureType :: (SNatI n) => Ty n -> TC n ()
+ensureType Kind = return ()
+ensureType ty = do
+  k <- inferType ty
+  unless (k == Kind) $ throwError "Not a type"
+
+-- | Infer the type of an expression.
+inferType :: (SNatI n) => Exp n -> TC n (Ty n)
+inferType (Var x) = do
+  (_, ty) <- get x
+  ensureType ty
+  return ty
+inferType Kind =
+  -- Kind is used internally to represent a well-formed
+  -- type, but should not be used otherwise.
+  throwError "Cannot type 'Kind'"
```

Comment:
`Kind` cannot be mentioned explicitly by the user.
We enforce this by refusing to type check `Kind`.